zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/ZkClient.java

```
@@ -62,7 +62,10 @@
   private static Logger LOG = LoggerFactory.getLogger(ZkClient.class);
 
   public static final int DEFAULT_OPERATION_TIMEOUT = Integer.MAX_VALUE;
-  public static final int DEFAULT_CONNECTION_TIMEOUT = 60 * 1000;
+  // Four times of DEFAULT_SESSION_TIMEOUT would give less 1/125 chance of initial connection
+  // to Zookeeper failure assuming one observer in the 5 observer Zookeeper Quorum is down
```

1. Please provide the formula instead of one set of input and result in the comment. Not all Helix deployment connects to a ZK which has 5 zk servers.
2. If we need 4 times of the session timeout, why hardcode here?
3. Since this configuration depends on the zk server count, I would prefer to specify this in the ZK_CONNECTION_TIMEOUT system property instead of modifying the default connection values. Note that this change might impact all the Helix users that do not specify the ZK_CONNECTION_TIMEOUT, and too long timeout may also cause problems.

---

For 3, indeed changing the property is a way. 

The fundamental problem is that if you want to fail fast or not. 

In helix case, only ZkHelixManager uses CONNECTION_TIMEOUT in connect(). If connect() does not succeed, it would throw exceptions and the application won't proceed. 

So it seems there is no downside to do the default change too.

---

@kaisun2000 

I also think that we should leave the constant as is and make it configurable in the system config.

Note that the HelixManager usage is _**the one you found**_. This is a public constant, and you don't know if it's being used elsewhere. So there is an issue of backward-compatibility if you change the value of the constant.